### PR TITLE
Support incidentDetection (#66)

### PR DESCRIPTION
**Pull request by HGlab01**
PR for https://github.com/iobroker-community-adapters/ioBroker.tado/issues/66 if it works for other users as well

### CODE REVIEW

**Comment by DutchmanNL** [comment]
Cool Thank you! Is it just a boolean value ?
We need to add the Attribute Definitions toi

https://github.com/iobroker-community-adapters/ioBroker.tado/blob/master/lib/state_attr.js

**Comment by HGlab01** [comment]
Yes, it is boolean:
"incidentDetection":{"supported":false,"enabled":false}

I need some tuturial session :-)
How to define these attributes if they are defined in a generic way in a "for" loop?

**Comment by DutchmanNL** [comment]
> I need some tuturial session :-)
> How to define these attributes if they are created in a generic way in a "for" loop?

good you ask and happy. to explain :)
I try to ensure for all my adapter the popper "State attributes" are set to avoid issue in upcoming JS-controller versions and proper (automated) integration (like role value.temperature)

To accomplish this, I build a generic state attribute function. Definitions. are stored in : 
https://github.com/iobroker-community-adapters/ioBroker.tado/blob/master/lib/state_attr.js

During state creations, the function : 

- receives the "name" which is provided by calling the function
https://github.com/iobroker-community-adapters/ioBroker.tado/blob/master/main.js#L1443
- create the states with definitions provided in statte_attr.js
https://github.com/iobroker-community-adapters/ioBroker.tado/blob/master/main.js#L1470-L1481
- When a value is not defined, I a fallback is called for "generic" definitions
https://github.com/iobroker-community-adapters/ioBroker.tado/blob/master/main.js#L1511-L1524

the proper definitions should be added. to state_attr.js

```
	'incidentDetection' : {
		name: 'Incident Detection',
		type: 'boolean',
		role: 'sensor.alarm ',
		unit:  ''
	},
```
(see for role definition : https://github.com/ioBroker/ioBroker/blob/master/doc/STATE_ROLES.md)

i just realized the create_state function in this adapter is build quite dirty, I will implement my proper updated version as used in wlanthermo which us much easier to read and understand :)

```
async doStateCreate(stateName, name, value) {

		// Strinfnify value if needed

		if (typeof (value) === 'object') { value = JSON.stringify(value); }

		// Try to get details from state lib, if not use defaults. throw warning if states is not known in attribute list
		const common = {};
		if (!stateAttr[name]) {
			const warnMessage = `State attribute definition missing for + ${name}`;
			if (warnMessages[name] !== warnMessage) {
				warnMessages[name] = warnMessage;
				console.warn(warnMessage);
				this.log.warn(warnMessage);

				// Send information to Sentry
				this.errorHandling(`[doStateCreate]`,warnMessage);
			}
		}

		common.name = stateAttr[name] !== undefined ? stateAttr[name].name || name : name;
		common.type = typeof (value);
		common.role = stateAttr[name] !== undefined ? stateAttr[name].role || 'state' : 'state';
		common.read = true;
		common.unit = stateAttr[name] !== undefined ? stateAttr[name].unit || '' : '';
		common.write = stateAttr[name] !== undefined ? stateAttr[name].write || false : false;

		if ((!this.createdStatesDetails[stateName])
			|| (this.createdStatesDetails[stateName]
				&& (
					common.name !== this.createdStatesDetails[stateName].name
					|| common.name !== this.createdStatesDetails[stateName].name
					|| common.type !== this.createdStatesDetails[stateName].type
					|| common.role !== this.createdStatesDetails[stateName].role
					|| common.read !== this.createdStatesDetails[stateName].read
					|| common.unit !== this.createdStatesDetails[stateName].unit
					|| common.write !== this.createdStatesDetails[stateName].write
				)
			)) {

			console.log(`An attribute has changed : ${stateName}`);

			await this.extendObjectAsync(stateName, {
				type: 'state',
				common
			});

		} else {
			// console.log(`Nothing changed do not update object`);
		}

		// Store current object definition to memory
		this.createdStatesDetails[stateName] = common;
		await this.setStateChangedAsync(stateName, { val: value, ack: true });
		common.write  && this.subscribeStates(stateName);
	}
```

**Comment by HGlab01** [comment]
Hi @DutchmanNL ,
thanx for these details and insights! Very appreciated.
Understood!

**Comment by DutchmanNL** [comment]
will you give it at try or should I update the files ?

**Comment by HGlab01** [comment]
I am now in the subway and will be late :-)
So, feel free to finalize the work...

**Comment by HGlab01** [comment]
> Cool Thank you! Is it just a boolean value ?
> We need to add the Attribute Definitions toi
> 
> https://github.com/iobroker-community-adapters/ioBroker.tado/blob/master/lib/state_attr.js

done ;-)

File: lib/state_attr.js [hunk 1]
```diff
@@ -338,8 +338,20 @@ const state_attrb = {
 		type: 'boolean',
 		role: 'info',
 		unit:  ''
-	}, 
+	},
+	'supported' : {
+		name: 'supported',
+		type: 'boolean',
+		role: 'info',
```
Comment (DutchmanNL) [inline lib/state_attr.js:345]: agree to have this as an info

File: lib/state_attr.js [hunk 2]
```diff
@@ -338,8 +338,20 @@ const state_attrb = {
...
+		unit:  ''
+	},
+	'enabled' : {
+		name: 'enabled',
+		type: 'boolean',
+		role: 'info',
```
Comment (DutchmanNL) [inline lib/state_attr.js:351]: role "info" does not exist in iobroker role definition overview : 
https://github.com/ioBroker/ioBroker/blob/master/doc/STATE_ROLES.md

Schermafbeelding 2020-09-29 om 09 28 53

I think it should. be an alarm type ?

Schermafbeelding 2020-09-29 om 09 29 41

**Review by DutchmanNL — CHANGES_REQUESTED** [review]
cool 👯 , please have a look at my review comments

File: lib/state_attr.js (hunk 1 above)
Comment (HGlab01) [reply lib/state_attr.js:345]: Then I would go with "state" for this

File: lib/state_attr.js (hunk 2 above)
Comment (HGlab01) [reply lib/state_attr.js:351]: and with indicator.alarm for that.
What do you think?
It's a little bit academically, isn't it?

**Comment by HGlab01** [comment]
I hope I used the GIT workflow correctly ;-)

**Comment by DutchmanNL** [comment]
> I hope I used the GIT workflow correctly ;-)

yes you did :)

**Pull request merged**